tools/xenpcid/xenpcid.c

```diff
@@ -0,0 +1,675 @@
+/*
+    Xenpcid daemon that acts as a server for the client in the libxl PCI
+
+    Copyright (C) 2021 EPAM Systems Inc.
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2.1 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define _GNU_SOURCE  // required for strchrnul()
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#include <libxenvchan.h>
+#include <xenpcid.h>
+#include <xenstore.h>
+
+#include <libxl.h>
+#include <libxl_json.h>
+#include <dirent.h>
+#include "pcid.h"
+
+#define BUFSIZE 4096
+/*
+ * TODO: Running this code in multi-threaded environment
+ * Now the code is designed so that only one request to the server
+ * from the client is made in one domain. In the future, it is necessary
+ * to take into account cases when from different domains there can be
+ * several requests from a client at the same time. Therefore, it will be
+ * necessary to regulate the multithreading of processes for global variables.
+ */
+char inbuf[BUFSIZE];
+char outbuf[BUFSIZE];
+int insiz = 0;
+int outsiz = 0;
+struct libxenvchan *ctrl;
+
+static void *pcid_zalloc(size_t size)
+{
+    void *ptr = calloc(size, 1);
+    if (!ptr) {
+        fprintf(stderr, "Memory allocation failed\n");
+        exit(1);
+    }
+    memset(ptr, 0, size);
+    return ptr;
+}
+
+static void free_pcid_obj_map(struct pcid__json_object *obj)
+{
+    if (obj->u.map)
+    {
+        unsigned int count;
+        struct flexarray *map;
+
+        map = obj->u.map;
+        count = map->count;
+        while (count--) {
+            struct pcid__json_map_node *node;
+            node = map->data[count];
```

Comment:
empty line is missing between variable declaration and code

Reply:
Done